api/src/main/java/org/apache/iceberg/metrics/FixedReservoirHistogram.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.metrics;
+
+import java.util.Arrays;
+import java.util.Random;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+/** A {@link Histogram} implementation with reservoir sampling. */
+public class FixedReservoirHistogram implements Histogram {
+  private final Random rand;
+  private final long[] measurements;
+  private int count;
+
+  public FixedReservoirHistogram(int reservoirSize) {
+    this.rand = new Random();
+    this.measurements = new long[reservoirSize];
+    this.count = 0;
+  }
+
+  @Override
+  public synchronized int count() {
+    return count;
+  }
+
+  @Override
+  public synchronized void update(long value) {
+    count += 1;
+    int index = count <= measurements.length ? count - 1 : rand.nextInt(count);
+    if (index < measurements.length) {
+      measurements[index] = value;
+    }
+  }
+
+  /**
+   * Naive algorithm for calculating variance:
+   * https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance
+   */
+  @Override
+  public Statistics statistics() {
+    int size = (int) Math.min(count, (long) measurements.length);
+    if (size == 0) {
+      return UniformWeightStatistics.EMPTY;
+    }
+
+    long[] values = new long[size];
```

Comment:
I think this should probably use just one `synchronized` here with `System.arraycopy` rather than locking for each value in the loop below.